### ISSUE TITLE
Override default type handling

### ISSUE THREAD

**Issue by ronnyek** [linq2db/linq2db#754]
Oracle doesnt or at least didnt have a boolean datatype available for actual column datatypes. It LOOKS as if the default  mapping for oracle uses a byte... but oracles recommendation for a long time anyway was to use a char field with a constraint of null or Y or N. 

Rather than change globally how oracle driver handles bool conversion... is there any mechanism to handle conversions to and from a given type for a given entity?

Or a way to override default handling in the oracle data provider, without having to maintain a fork of the oracle provider?

I searched the wiki as well as looked into some really old references talking abotu expressions (that look to achieve different functionality)

### Environment details
linq2db version: *1.8.3-rc958*  
Database Server: *E.g. Oracle 12c*

**Comment by RoyC-Retail-Assist** [comment]
I had a similar issue with a DB2 database.  I created a BoolYN class that is convertible to/from bool.  This is then used as the type for the column property with a Column attribute specifying the storage type.

I can sort a Gist if you are interested.

**Comment by ili** [comment]
What about 
```cs
var schema = MappingSchema.Default()
schema.SetConverter<bool, DataParameter>(...)
schema.SetConverter<DataParameter, bool>(...)
schema.SetConverter<TypeYouChoosenToStoreBool, bool>(...)
```

**Comment by ronnyek** [comment]
so this actually doesn't appear to work... I was for a while implementing my own dedicated oracle provider, but the differences were so minimal, I felt like I needed to revert back to the built in stuff (for future compatibility). 

The examples above for converting bool to data parameter dont work, as they require a name parameter, which would probably be invalid no matter what string I would provide.

I was overriding setParameter, and checking for bool and setting Y/N accordingly. Is there anyway to do the equivalent without having to create an inherited provider?

**Comment by ronnyek** [comment]
It seems I was needing to use setConvertExpression like below.

` 
var schema = MappingSchema.Default;
schema.SetConvertExpression<bool, DataParameter>(_ => DataParameter.Char(null, _ ? 'Y' : 'N'));
schema.SetConvertExpression<char, bool>(_ => _ == 'Y');
schema.SetConvertExpression<string, bool>(_ => _.Trim() == "Y");
schema.SetConvertExpression<bool?, DataParameter>(_ => _.HasValue ? DataParameter.Char(null,_.Value  'Y' : 'N') : new DataParameter(null, null));
schema.SetConvertExpression<char?, bool?>(_ => _.HasValue ? _.Value == 'Y' : default(bool?));`

**Comment by sdanyliv** [comment]
@ronnyek, we can fix our provider easily if you will provide us with schema and which results you are expected.
Yes, we are not Oracle guys (but looks like we become experts in many databases) and external help is very appreciated.
If you provide PR with tests - everything will be done smoothly.

**Comment by ronnyek** [comment]
Well the problem here is I'm not sure whats currently there in the default oracle provider is wrong. Oracle didn't have a real bit type you could  use outside of PLSQL, so at the time the recommendation was to use a char field with constraints to either allow Y or N. 

I believe they now allow support for bit fields, but wouldnt be surprised if people came up with other mechanisms to support bools where there was no db support.

After running my tests with the SetConvertExpression stuff I posted above, I think it actually does work how I expect it to, and can use an otherwise pure oracle data provider.

Now just to help get linq2db updated to latest oracle .net core driver and whatever I can do to help with nuget packaging of it! 

@sdanyliv I think we can re-close this one for now. Appreciate the time

**Comment by ronnyek** [comment]
I'm so sorry to keep going back and forth...  the code in https://github.com/linq2db/linq2db/issues/754#issuecomment-398513687
as well as trying
`db.MappingSchema.SetConvertExpression<bool?, DataParameter>(_ => _ != null ? DataParameter.Char(null, _.HasValue && _.Value ? 'Y' : 'N') : new DataParameter(null, System.DBNull.Value));`

Seems to actually work well in almost every situation, except when I'm passing in a nullable bool as null. 
I'm getting a System.NullReference exception on line 164 of QueryRunner.cs where it tries to call this:
```
var value = p.Accessor(expression, parameters);
```

As far as telling linq2db how to convert types into values is this still the correct place to do it, or should I be doing something else? Again, this is one of those situations where I think there isn't a great default for this, so I suspect people will run into this with oracle. Some people use numbers and 1/0, some people use chars and use Y/N.

**Comment by sdanyliv** [comment]
@ronnyek, will check tomorrow.

**Comment by sdanyliv** [comment]
@ronnyek, I have prepared fix for that 43b6d9602ad0faed93151da0b94e1aff7cdb4ff8

**Comment by MaceWindu** [comment]
Looks like it is already addressed, please reopen/create new issue if you still have issues with it
